### ISSUE TITLE
Make cross_shard_tx adaptively change its block production time

### ISSUE THREAD

**Issue by SkidanovAlex** [nearprotocol/nearcore#3014]
The usefulness and flakiness of `cross_shard_tx` depends heavily on the block production time.

If the block production time is too high, there isn't many forks, and the test passes easily without testing any complex scenarios.

If the block production time is too low, with doomslug disabled it often degrades into a situation in which exactly half of the validators are on one fork, and exactly half are on the other, and they both produce blocks faster than they can sync to the other chain.

One way to address it is to dynamically change the block production time.

To know if it's too high we can maintain the ratio of total number of blocks to the length of the longest chain (where the longest chain is defined as the number of blocks, not height).

To know if it's too little we can maintain the depth of the longest fork, and if it is too deep, bring the block prod delay slightly higher, but no higher than say 500ms.

https://github.com/nearprotocol/nearcore/issues/3014

**Comment by pmnoxx** [comment]
@SkidanovAlex Thanks for the explanation. I see that  doomslug.rs::process_timer method defines how we increase `height` based on current time. `Dumslug.timer.get_delay` is used to compute the amount of time spend on given height.

Instead of dynamically adjusting the block production time wouldn't it be simpler to do a binary search, run test multiple times. We could measure the number of forks, length of canonical chain after each run, etc. This would allow us to binary search the optimal block_production_time. This would increase the amount of time this test needs to run by a factor of 5-10.

However, if we really want to do the adjustment dynamically. We need to guarantee that we change the block production time for all ClientActors while they are at the same height. We want to avoid a situation when some ClientActors timers get desynchronized and they expect blocks at given heigh at different time. 

There are a few possible approaches:
1) direct call all client actors - won't work
In actix there is no way to call method directly on ClientActors

2) send message to all client actors using actix.
`ChangeProductionDelay{ production_delay: u32, starting_height: u32 }.`
We could send a message, which tells client actors to change production delay at given height. Although, we would need to some reasonable delay. It may be tricky to configure this correctly. This would also require us to modify the production code, change doomslug implementation to do this. This approach may be hard to maintain, because block production time is used multiple times throughout the code.

I don't see any clean solution to this problem. Running test multiple timeswith a binary search seems like a much cleaner solution to me. What do you think?

**Comment by pmnoxx** [comment]
I see that this test takes 88 minutes, so maybe it's probably not good idea to make it run even longer.

Instead of changing block production time to cause forks to happen, we could change the test and add a packet loss in the network mock. 
If the number of forks is too low, we can increase it, otherwise decrease is.

**Comment by SkidanovAlex** [comment]
**Re: Updating clients simultaneously**
Clients generally should continue operating even if briefly they have had inconsistent delays (the very same delays might be natural due to network conditions, and doomslug for liveness assumes that there could be arbitrarily long network delays, as long as they are followed by a sufficiently long period of bounded delays).

Re: packet drop
The basic `cross_shard_tx` is not dropping packets, and this provides additional coverage compared to the version of the test that does drop packets (`cross_shard_tx_drop_chunks`). Ideally I would not want to remove the version that doesn't drop packets.

**Re: binary search**
This I think is a great idea, if we do it with a twist. If we can make the block production time a command line argument (I don't know if it is possible with Rust test infra), then we can make the following adjustment:

1. Still maintain the statistics of ratio of blocks to longest chain (low ratio => block prod is too high) and length of the longest forks (too high => block prod is too low)
2. If we see that the block prod is too low or too high, instead of dynamically adjusting the block prod, fail the test.
3. Given nightly runners should have consistent hardware, we just manually tune the block prod time for the nightly runs. If people want to run the test locally, they start with the value from the night run, and if it's too low or too high for their setup, they adjust it. They only need to do it once for their setup as well.

**Comment by pmnoxx** [comment]
> **Re: Updating clients simultaneously**
> Clients generally should continue operating even if briefly they have had inconsistent delays (the very same delays might be natural due to network conditions, and doomslug for liveness assumes that there could be arbitrarily long network delays, as long as they are followed by a sufficiently long period of bounded delays).

The issue here is with the timer that doomslug uses. It's defined as `{ started (time stamp of block to be produced at given height), height }`. It's basically a mapping from current time to expected height of the current block.

There is a possible race condition, that you can check out below:

Let's assume current block production time is 100 ms.

ClientActor 1: 
at time X ... X + 100ms, wants to produce blocks at height H, 
it gets message to send block production time to 200ms.
At time X + 100ms ... X + 300ms, it expects to produce block H + 1.
At time X + 300ms ... X + 500ms, it expects to produce block H + 2.

ClientActor 2:
 at time X ... X + 100ms, wants to produce blocks at height H, 
 at time X + 100ms... X + 200ms, wants to produce blocks at height H + 1,  
it gets message to send block production time to 200ms.
At time X + 200ms ... X + 400ms, it expects to produce block H + 2.

Now both client actors expect block H + 2 to be produced at different intervals, if we continue changing block production time, the interval at which they expect block to be produced may drift apart.

> **Re: binary search**
> This I think is a great idea, if we do it with a twist. If we can make the block production time a command line argument (I don't know if it is possible with Rust test infra), then we can make the following adjustment:
> 
> 1. Still maintain the statistics of ratio of blocks to longest chain (low ratio => block prod is too high) and length of the longest forks (too high => block prod is too low)
> 2. If we see that the block prod is too low or too high, instead of dynamically adjusting the block prod, fail the test.
> 3. Given nightly runners should have consistent hardware, we just manually tune the block prod time for the nightly runs. If people want to run the test locally, they start with the value from the night run, and if it's too low or too high for their setup, they adjust it. They only need to do it once for their setup as well
Like you mentioned, one way of doing this is to do binary search by the testing infrastructure, I'm not sure if it's possible.

I was thinking about doing the binary search within a single test. We run a single run of the test, get statistics, stop actix, then adjust the delay, run it another time, etc.

**Comment by SkidanovAlex** [comment]
> Now both client actors expect block H + 2 to be produced at different intervals, if we continue changing block production time, the interval at which they expect block to be produced may drift apart.

The timer however resets each time a block is received (and the head is updated). If we only change the block production time when statistics change, it must be that at least some blocks were produced since the last block was received.

Also, in doomslug such a discrepancy can result in stalling only if someone intentionally tries to take advantage of it (which the test doesn't). Otherwise even if clients permanently had wildly different block production times, the blocks would still get produced, because there will be moments when a particular block producer has enough endoresments / skips to produce a block.

> I was thinking about doing the binary search within a single test. We run a single run of the test, get statistics, stop actix, then adjust the delay, run it another time, etc.

Yeah, but I figured that you didn't like the idea because that takes a lot of time. Running in parallel doesn't help too much, we can may be run 4 in parallel on an 8 core machien without drastically affecting the test itself.

But running binary search once manually sounds like a good idea. In the past whenever I found a good block prod time locally (after changes that made the test behave differently), that block prod time generally worked good until we changed the server drastically again. If the test would fail with an informative error mesasge whenever that happens (which it would be able to detect if it tracks forkfullness / length of forks), I think it's good enough of a solution for now.
